File: typescript/packages/core/tests/observability.test.ts
/**
 * Unit tests for the canonical ObservabilityProvider (spore-core issue #12).
 *
 * Mirrors `rust/crates/spore-core/src/observability.rs#tests` rule-for-rule.
 */

import { describe, expect, it } from "vitest";

import {
  guideRegistry,
  memory,
  middleware,
  observability,
  sensor,
  SessionId,
  TaskId,
} from "../src/index.js";

const { InMemoryObservabilityProvider, SpanId, PricingTable } = observability;
type SpanBase = observability.SpanBase;
type TurnSpan = observability.TurnSpan;
type ToolCallSpan = observability.ToolCallSpan;
type SensorSpan = observability.SensorSpan;
type ContextSpan = observability.ContextSpan;
type ContextOperation = observability.ContextOperation;
type MiddlewareSpan = observability.MiddlewareSpan;

const { Timestamp } = memory;
const { SensorId } = sensor;

function sid(s: string): SessionId {
  return SessionId.of(s);
}
function tid(s: string): TaskId {
  return TaskId.of(s);
}
function ts(s: string): memory.Timestamp {
  return Timestamp.of(s);
}

function turnSpan(
  session: string,
  spanId: string,
  turn: number,
  input: number,
  output: number,
): TurnSpan {
  const base: SpanBase = {
    span_id: SpanId.of(spanId),
    parent_span_id: null,
    session_id: sid(session),
    task_id: tid("t1"),
    kind: "turn",
    started_at: ts("2026-05-16T00:00:00Z"),
    ended_at: ts("2026-05-16T00:00:01Z"),
    duration_ms: 1000,
    status: { kind: "ok" },
  };
  return {
    base,
    turn_number: turn,
    input_tokens: input,
    output_tokens: output,
    cache_read_tokens: null,
    cache_write_tokens: null,
    cost_usd: 0,
    stop_reason: "end_turn",
    tool_calls_requested: 0,
  };
}

describe("InMemoryObservabilityProvider", () => {
  // ── Rule: emitTurn is fire-and-forget (sync) and span is queryable ────────
  it("emit_turn recorded and metrics aggregate", async () => {
    const obs = new InMemoryObservabilityProvider();
    obs.emitTurn(turnSpan("s1", "sp1", 1, 100, 50));
    obs.emitTurn(turnSpan("s1", "sp2", 2, 200, 80));
    obs.setSessionOutcome(sid("s1"), { kind: "success" });
    const m = await obs.getSessionMetrics(sid("s1"));
    expect(m).toBeDefined();
    expect(m!.total_turns).toBe(2);
    expect(m!.total_input_tokens).toBe(300);
    expect(m!.total_output_tokens).toBe(130);
    expect(m!.outcome).toEqual({ kind: "success" });
  });

  // ── Rule: emit_tool_call counted in metrics ───────────────────────────────
  it("emit_tool_call counted in metrics", async () => {
    const obs = new InMemoryObservabilityProvider();
    obs.emitTurn(turnSpan("s1", "t1", 1, 10, 5));
    const base: SpanBase = {
      span_id: SpanId.of("tc1"),
      parent_span_id: null,
      session_id: sid("s1"),
      task_id: tid("t1"),
      kind: "tool_call",
      started_at: ts("2026-05-16T00:00:00Z"),
      ended_at: ts("2026-05-16T00:00:00Z"),
      duration_ms: 250,
      status: { kind: "ok" },
    };
    const tc: ToolCallSpan = {
      base,
      tool_name: "shell",
      call_id: "c1",
      parameters_size_bytes: 12,
      output_size_bytes: 42,
      truncated: false,
      sandbox_mode: "workspace_scoped",
      sandbox_violations: [],
    };
    obs.emitToolCall(tc);
    const m = await obs.getSessionMetrics(sid("s1"));
    expect(m!.tool_calls).toBe(1);
    expect(m!.total_duration_ms).toBe(1250);
  });

  // ── Rule: sensor metrics — fires and halts ────────────────────────────────
  it("sensor metrics count fires and halts", async () => {
    const obs = new InMemoryObservabilityProvider();
    obs.emitTurn(turnSpan("s1", "t1", 1, 10, 5));
    const mk = (id: string, fired: boolean, outcome: sensor.SensorOutcome): SensorSpan => ({
      base: {
        span_id: SpanId.of(id),
        parent_span_id: null,
        session_id: sid("s1"),
        task_id: tid("t1"),
        kind: "sensor_evaluation",
        started_at: ts("2026-05-16T00:00:00Z"),
        ended_at: ts("2026-05-16T00:00:00Z"),
        duration_ms: 1,
        status: { kind: "ok" },
      },
      sensor_id: SensorId.of("lint"),
      sensor_kind: "computational",
      trigger: { kind: "post_turn" },
      outcome,
      fired,
    });
    obs.emitSensor(mk("sn1", true, "warn"));
    obs.emitSensor(mk("sn2", true, "halt"));
    obs.emitSensor(mk("sn3", false, "pass"));
    const m = await obs.getSessionMetrics(sid("s1"));
    expect(m!.sensor_fires).toBe(2);
    expect(m!.sensor_halts).toBe(1);
  });

  // ── Rule: compaction counted ──────────────────────────────────────────────
  it("compaction counted in metrics", async () => {
    const obs = new InMemoryObservabilityProvider();
    obs.emitTurn(turnSpan("s1", "t1", 1, 100, 50));
    const mkCtx = (op: ContextOperation): ContextSpan => ({
      base: {
        span_id: SpanId.of(op.kind === "compaction" ? "c1" : "c2"),
        parent_span_id: null,
        session_id: sid("s1"),
        task_id: tid("t1"),
        kind: op.kind === "compaction" ? "compaction" : "context_assembly",
        started_at: ts("2026-05-16T00:00:00Z"),
        ended_at: ts("2026-05-16T00:00:00Z"),
        duration_ms: 1,
        status: { kind: "ok" },
      },
      operation: op,
      tokens_before: 10_000,
      tokens_after: 5_000,
      utilization_before: 0.9,
      utilization_after: 0.5,
    });
    obs.emitContext(mkCtx({ kind: "compaction", messages_removed: 5, tokens_reclaimed: 5_000 }));
    obs.emitContext(
      mkCtx({
        kind: "assembly",
        guides_loaded: 2,
        memory_items_loaded: 3,
        tools_loaded: 5,
      }),
    );
    const m = await obs.getSessionMetrics(sid("s1"));
    expect(m!.compactions).toBe(1);
  });

  // ── Rule: pricing table computes cost_usd ─────────────────────────────────
  it("pricing table computes cost", () => {
    const table: observability.PricingTable = {
      input_per_million: 3.0,
      output_per_million: 15.0,
      cache_read_per_million: 0.3,
      cache_write_per_million: 3.75,
    };
    const cost = PricingTable.costFor(table, 1_000_000, 1_000_000, 1_000_000, 1_000_000);
    // 3 + 15 + 0.3 + 3.75 = 22.05
    expect(Math.abs(cost - 22.05)).toBeLessThan(1e-9);
  });

  it("pricing table DEFAULT is zero", () => {
    const cost = PricingTable.costFor(PricingTable.DEFAULT, 1000, 1000, 1000, 1000);
    expect(cost).toBe(0);
  });

  // ── Rule: flushSession idempotent + spans remain queryable ────────────────
  it("flushSession is idempotent and spans remain queryable", async () => {
    const obs = new InMemoryObservabilityProvider();
    obs.emitTurn(turnSpan("s1", "t1", 1, 10, 5));
    await obs.flushSession(sid("s1"));
    await obs.flushSession(sid("s1")); // second flush is a no-op
    const m = await obs.getSessionMetrics(sid("s1"));
    expect(m!.total_turns).toBe(1);
    const trace = await obs.getTrace(sid("s1"));
    expect(trace.length).toBe(1);
  });

  // ── Rule: getTrace returns spans in insertion order; parent linkage kept ──
  it("getTrace preserves insertion order and parent linkage", async () => {
    const obs = new InMemoryObservabilityProvider();
    obs.emitTurn(turnSpan("s1", "sp1", 1, 10, 5));
    const tc: ToolCallSpan = {
      base: {
        span_id: SpanId.of("sp2"),
        parent_span_id: SpanId.of("sp1"),
        session_id: sid("s1"),
        task_id: tid("t1"),
        kind: "tool_call",
        started_at: ts("2026-05-16T00:00:00Z"),
        ended_at: ts("2026-05-16T00:00:00Z"),
        duration_ms: 1,
        status: { kind: "ok" },
      },
      tool_name: "shell",
      call_id: "c1",
      parameters_size_bytes: 0,
      output_size_bytes: 0,
      truncated: false,
      sandbox_mode: "none",
      sandbox_violations: [],
    };
    obs.emitToolCall(tc);
    const trace = await obs.getTrace(sid("s1"));
    expect(trace.length).toBe(2);
    expect(trace[0].base.span_id.asString()).toBe("sp1");
    expect(trace[1].base.span_id.asString()).toBe("sp2");
    expect(trace[1].base.parent_span_id?.asString()).toBe("sp1");
  });

  // ── Rule: middleware span recorded with hook + decision ───────────────────
  it("middleware span recorded in trace", async () => {
    const obs = new InMemoryObservabilityProvider();
    const span: MiddlewareSpan = {
      base: {
        span_id: SpanId.of("mw1"),
        parent_span_id: null,
        session_id: sid("s1"),
        task_id: tid("t1"),
        kind: "middleware_hook",
        started_at: ts("2026-05-16T00:00:00Z"),
        ended_at: ts("2026-05-16T00:00:00Z"),
        duration_ms: 0,
        status: { kind: "ok" },
      },
      hook: "before_turn" as middleware.HookPoint,
      decision: { kind: "continue" } as middleware.MiddlewareDecision,
    };
    obs.emitMiddleware(span);
    const trace = await obs.getTrace(sid("s1"));
    expect(trace.length).toBe(1);
    expect(trace[0].base.kind).toBe("middleware_hook");
  });

  // ── Rule: getSessions filters by outcome ──────────────────────────────────
  it("getSessions filters by outcome", async () => {
    const obs = new InMemoryObservabilityProvider();
    obs.emitTurn(turnSpan("good", "sp1", 1, 10, 5));
    obs.emitTurn(turnSpan("bad", "sp2", 1, 10, 5));
    obs.setSessionOutcome(sid("good"), { kind: "success" });
    obs.setSessionOutcome(sid("bad"), { kind: "failure", reason: "x" });
    const success = await obs.getSessions(ts("2026-05-16T00:00:00Z"), undefined, {
      kind: "success",
    });
    expect(success.length).toBe(1);
    expect(success[0].session_id.asString()).toBe("good");
  });

  // ── Rule: getSessions filters by since timestamp ──────────────────────────
  it("getSessions filters by since", async () => {
    const obs = new InMemoryObservabilityProvider();
    const early = turnSpan("old", "sp1", 1, 10, 5);
    early.base = { ...early.base, started_at: ts("2026-01-01T00:00:00Z") };
    obs.emitTurn(early);
    obs.emitTurn(turnSpan("new", "sp2", 1, 10, 5));
    const recent = await obs.getSessions(ts("2026-05-15T00:00:00Z"));
    const ids = recent.map((m) => m.session_id.asString());
    expect(ids).toContain("new");
    expect(ids).not.toContain("old");
  });

  // ── Rule: recordGuidesUsed surfaces on metrics ────────────────────────────
  it("guides_used surfaces on metrics", async () => {
    const obs = new InMemoryObservabilityProvider();
    obs.emitTurn(turnSpan("s1", "sp1", 1, 10, 5));
    const g = guideRegistry.GuideId.of("g1");
    obs.recordGuidesUsed(sid("s1"), [g]);
    const m = await obs.getSessionMetrics(sid("s1"));
    expect(m!.guides_used.map((x) => x.asString())).toEqual(["g1"]);
  });

  // ── Rule: passive observer — no harness-mutating method on interface ──────
  // (type-level: ObservabilityProvider has no method that returns a value the
  //  harness uses to alter behavior. Asserted here by structural usage.)
  it("ObservabilityProvider can be used via the interface alone", async () => {
    const provider: observability.ObservabilityProvider = new InMemoryObservabilityProvider();
    provider.emitTurn(turnSpan("s1", "sp1", 1, 10, 5));
    const m = await provider.getSessionMetrics(sid("s1"));
    expect(m!.total_turns).toBe(1);
  });

  // ── SpanBase helpers ──────────────────────────────────────────────────────
  it("new root and child spans", () => {
    const root = observability.newRootSpanBase(
      SpanId.of("r"),
      sid("s"),
      tid("t"),
      "session",
      ts("2026-05-16T00:00:00Z"),
    );
    const child = observability.newChildSpanBase(
      SpanId.of("c"),
      root,
      "turn",
      ts("2026-05-16T00:00:01Z"),
    );
    expect(child.parent_span_id?.asString()).toBe("r");
    expect(child.session_id.asString()).toBe("s");
  });
});

File: go/spore-core/observability/observability_test.go
package observability

import (
	"context"
	"math"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/guideregistry"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/memory"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/middleware"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/sensor"
)

// ─── helpers ────────────────────────────────────────────────────────────────

func ts(s string) Timestamp  { return memory.Timestamp(s) }
func sid(s string) SessionID { return sporecore.SessionID(s) }
func tid(s string) TaskID    { return sporecore.TaskID(s) }

func turnSpan(session, spanID string, turn, input, output uint32) TurnSpan {
	return TurnSpan{
		Base: SpanBase{
			SpanID:     SpanID(spanID),
			SessionID:  sid(session),
			TaskID:     tid("t1"),
			Kind:       SpanKindTurn,
			StartedAt:  ts("2026-05-16T00:00:00Z"),
			EndedAt:    ts("2026-05-16T00:00:01Z"),
			DurationMs: 1000,
			Status:     NewStatusOk(),
		},
		TurnNumber:   turn,
		InputTokens:  input,
		OutputTokens: output,
		StopReason:   sporecore.StopEndTurn,
	}
}

// ─── Rule: EmitTurn is fire-and-forget; metrics aggregate ───────────────────

func TestEmitTurnRecordedAndMetricsAggregate(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	obs.EmitTurn(turnSpan("s1", "sp1", 1, 100, 50))
	obs.EmitTurn(turnSpan("s1", "sp2", 2, 200, 80))
	obs.SetSessionOutcome(sid("s1"), guideregistry.NewOutcomeSuccess())

	m, err := obs.GetSessionMetrics(context.Background(), sid("s1"))
	if err != nil || m == nil {
		t.Fatalf("metrics: err=%v m=%v", err, m)
	}
	if m.TotalTurns != 2 || m.TotalInputTokens != 300 || m.TotalOutputTokens != 130 {
		t.Fatalf("aggregates wrong: %+v", m)
	}
	if m.Outcome.Kind != guideregistry.OutcomeKindSuccess {
		t.Fatalf("outcome: %v", m.Outcome)
	}
}

// ─── Rule: EmitToolCall counted in metrics; duration accumulates ────────────

func TestEmitToolCallCountedInMetrics(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	obs.EmitTurn(turnSpan("s1", "t1", 1, 10, 5))
	obs.EmitToolCall(ToolCallSpan{
		Base: SpanBase{
			SpanID:     "tc1",
			SessionID:  sid("s1"),
			TaskID:     tid("t1"),
			Kind:       SpanKindToolCall,
			StartedAt:  ts("2026-05-16T00:00:00Z"),
			EndedAt:    ts("2026-05-16T00:00:00Z"),
			DurationMs: 250,
			Status:     NewStatusOk(),
		},
		ToolName:    "shell",
		CallID:      "c1",
		SandboxMode: "workspace_scoped",
	})
	m, _ := obs.GetSessionMetrics(context.Background(), sid("s1"))
	if m.ToolCalls != 1 {
		t.Fatalf("tool_calls=%d", m.ToolCalls)
	}
	if m.TotalDurationMs != 1250 {
		t.Fatalf("duration=%d", m.TotalDurationMs)
	}
}

// ─── Rule: Sensor metrics count fires and halts ─────────────────────────────

func TestSensorMetricsCountFiresAndHalts(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	obs.EmitTurn(turnSpan("s1", "t1", 1, 10, 5))
	mk := func(id string, fired bool, outcome SensorOutcome) SensorSpan {
		return SensorSpan{
			Base: SpanBase{
				SpanID:    SpanID(id),
				SessionID: sid("s1"),
				TaskID:    tid("t1"),
				Kind:      SpanKindSensorEvaluation,
				StartedAt: ts("2026-05-16T00:00:00Z"),
				EndedAt:   ts("2026-05-16T00:00:00Z"),
				Status:    NewStatusOk(),
			},
			SensorID:   sensor.SensorID("lint"),
			SensorKind: sensor.SensorKindComputational,
			Trigger:    sensor.NewTriggerPostTurn(),
			Outcome:    outcome,
			Fired:      fired,
		}
	}
	obs.EmitSensor(mk("sn1", true, sensor.OutcomeWarn))
	obs.EmitSensor(mk("sn2", true, sensor.OutcomeHalt))
	obs.EmitSensor(mk("sn3", false, sensor.OutcomePass))
	m, _ := obs.GetSessionMetrics(context.Background(), sid("s1"))
	if m.SensorFires != 2 {
		t.Fatalf("fires=%d", m.SensorFires)
	}
	if m.SensorHalts != 1 {
		t.Fatalf("halts=%d", m.SensorHalts)
	}
}

// ─── Rule: Compactions counted in metrics ───────────────────────────────────

func TestCompactionCountedInMetrics(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	obs.EmitTurn(turnSpan("s1", "t1", 1, 100, 50))
	mkCtx := func(op ContextOperation) ContextSpan {
		return ContextSpan{
			Base: SpanBase{
				SpanID:    "c1",
				SessionID: sid("s1"),
				TaskID:    tid("t1"),
				Kind:      SpanKindCompaction,
				StartedAt: ts("2026-05-16T00:00:00Z"),
				EndedAt:   ts("2026-05-16T00:00:00Z"),
				Status:    NewStatusOk(),
			},
			Operation:    op,
			TokensBefore: 10000, TokensAfter: 5000,
			UtilizationBefore: 0.9, UtilizationAfter: 0.5,
		}
	}
	obs.EmitContext(mkCtx(NewContextOpCompaction(5, 5000)))
	obs.EmitContext(mkCtx(NewContextOpAssembly(2, 3, 5)))
	m, _ := obs.GetSessionMetrics(context.Background(), sid("s1"))
	if m.Compactions != 1 {
		t.Fatalf("compactions=%d", m.Compactions)
	}
}

// ─── Rule: PricingTable computes cost; DefaultPricing returns zero ──────────

func TestPricingTableComputesCost(t *testing.T) {
	table := PricingTable{
		InputPerMillion:      3.0,
		OutputPerMillion:     15.0,
		CacheReadPerMillion:  0.3,
		CacheWritePerMillion: 3.75,
	}
	one := uint32(1_000_000)
	cost := table.CostFor(one, one, &one, &one)
	// 3 + 15 + 0.3 + 3.75 = 22.05
	if math.Abs(cost-22.05) > 1e-9 {
		t.Fatalf("cost=%v want 22.05", cost)
	}
}

func TestDefaultPricingIsZero(t *testing.T) {
	one := uint32(1000)
	cost := DefaultPricing().CostFor(one, one, &one, &one)
	if cost != 0 {
		t.Fatalf("default cost=%v want 0", cost)
	}
}

func TestPricingTableHandlesNilCacheTokens(t *testing.T) {
	table := PricingTable{InputPerMillion: 3.0, OutputPerMillion: 15.0}
	// 3 * 1 + 15 * 1 = 18 (no cache tokens contributed)
	cost := table.CostFor(1_000_000, 1_000_000, nil, nil)
	if math.Abs(cost-18.0) > 1e-9 {
		t.Fatalf("cost=%v want 18", cost)
	}
}

// ─── Rule: FlushSession is idempotent — spans remain queryable ──────────────

func TestFlushSessionIsIdempotent(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	obs.EmitTurn(turnSpan("s1", "t1", 1, 10, 5))
	ctx := context.Background()
	if err := obs.FlushSession(ctx, sid("s1")); err != nil {
		t.Fatal(err)
	}
	if err := obs.FlushSession(ctx, sid("s1")); err != nil {
		t.Fatal(err)
	}
	m, _ := obs.GetSessionMetrics(ctx, sid("s1"))
	if m == nil || m.TotalTurns != 1 {
		t.Fatalf("metrics post-flush: %+v", m)
	}
}

// ─── Rule: GetTrace returns spans in insertion order; ParentSpanID kept ─────

func TestGetTracePreservesInsertionOrder(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	obs.EmitTurn(turnSpan("s1", "sp1", 1, 10, 5))
	parent := SpanID("sp1")
	obs.EmitToolCall(ToolCallSpan{
		Base: SpanBase{
			SpanID:       "sp2",
			ParentSpanID: &parent,
			SessionID:    sid("s1"),
			TaskID:       tid("t1"),
			Kind:         SpanKindToolCall,
			StartedAt:    ts("2026-05-16T00:00:00Z"),
			EndedAt:      ts("2026-05-16T00:00:00Z"),
			Status:       NewStatusOk(),
		},
		ToolName: "shell", CallID: "c1", SandboxMode: "none",
	})
	trace, err := obs.GetTrace(context.Background(), sid("s1"))
	if err != nil {
		t.Fatal(err)
	}
	if len(trace) != 2 {
		t.Fatalf("trace len=%d", len(trace))
	}
	if trace[0].GetBase().SpanID != "sp1" {
		t.Fatalf("ordering[0]=%v", trace[0].GetBase().SpanID)
	}
	if trace[1].GetBase().SpanID != "sp2" {
		t.Fatalf("ordering[1]=%v", trace[1].GetBase().SpanID)
	}
	if trace[1].GetBase().ParentSpanID == nil || *trace[1].GetBase().ParentSpanID != "sp1" {
		t.Fatalf("parent linkage lost: %+v", trace[1].GetBase().ParentSpanID)
	}
}

// ─── Rule: Middleware spans recorded with hook & decision ───────────────────

func TestMiddlewareSpanRecordedInTrace(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	obs.EmitMiddleware(MiddlewareSpan{
		Base: SpanBase{
			SpanID:    "mw1",
			SessionID: sid("s1"),
			TaskID:    tid("t1"),
			Kind:      SpanKindMiddlewareHook,
			StartedAt: ts("2026-05-16T00:00:00Z"),
			EndedAt:   ts("2026-05-16T00:00:00Z"),
			Status:    NewStatusOk(),
		},
		Hook:     middleware.HookBeforeTurn,
		Decision: middleware.DecisionContinueVal(),
	})
	trace, _ := obs.GetTrace(context.Background(), sid("s1"))
	if len(trace) != 1 {
		t.Fatalf("trace len=%d", len(trace))
	}
	if trace[0].GetBase().Kind != SpanKindMiddlewareHook {
		t.Fatalf("kind=%v", trace[0].GetBase().Kind)
	}
}

// ─── Rule: GetSessions filters by outcome ───────────────────────────────────

func TestGetSessionsFiltersByOutcome(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	obs.EmitTurn(turnSpan("good", "sp1", 1, 10, 5))
	obs.EmitTurn(turnSpan("bad", "sp2", 1, 10, 5))
	obs.SetSessionOutcome(sid("good"), guideregistry.NewOutcomeSuccess())
	obs.SetSessionOutcome(sid("bad"), guideregistry.NewOutcomeFailure("x"))
	want := guideregistry.NewOutcomeSuccess()
	got, err := obs.GetSessions(context.Background(), ts("2026-05-16T00:00:00Z"), nil, &want)
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 1 {
		t.Fatalf("len=%d", len(got))
	}
	if got[0].SessionID != sid("good") {
		t.Fatalf("session=%v", got[0].SessionID)
	}
}

// ─── Rule: GetSessions filters by since timestamp ───────────────────────────

func TestGetSessionsFiltersBySince(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	old := turnSpan("old", "sp1", 1, 10, 5)
	old.Base.StartedAt = ts("2026-01-01T00:00:00Z")
	obs.EmitTurn(old)
	obs.EmitTurn(turnSpan("new", "sp2", 1, 10, 5))
	got, err := obs.GetSessions(context.Background(), ts("2026-05-15T00:00:00Z"), nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	ids := map[SessionID]bool{}
	for _, m := range got {
		ids[m.SessionID] = true
	}
	if !ids[sid("new")] {
		t.Fatalf("missing new: %+v", ids)
	}
	if ids[sid("old")] {
		t.Fatalf("should not contain old: %+v", ids)
	}
}

// ─── Rule: passive observer — interface has no behavior-mutating methods ────
//
// Compile-time check: the ObservabilityProvider interface declares only
// EmitX (no return), FlushSession, and GetX accessors. There is no method
// that could change a TurnResult or ToolOutput. We also assert thread
// safety by exercising concurrent emits.

func TestObservabilityProviderIsConcurrencySafe(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	done := make(chan struct{}, 4)
	emit := func(prefix string) {
		for i := 0; i < 25; i++ {
			obs.EmitTurn(turnSpan("s1", prefix+string(rune('a'+i%26)), uint32(i), 1, 1))
		}
		done <- struct{}{}
	}
	go emit("a")
	go emit("b")
	go emit("c")
	go emit("d")
	for i := 0; i < 4; i++ {
		<-done
	}
	m, _ := obs.GetSessionMetrics(context.Background(), sid("s1"))
	if m == nil || m.TotalTurns != 100 {
		t.Fatalf("turns=%v", m)
	}
}

// ─── SpanBase helpers ───────────────────────────────────────────────────────

func TestNewRootAndNewChild(t *testing.T) {
	root := NewRoot(SpanID("r"), sid("s"), tid("t"), SpanKindSession, ts("2026-05-16T00:00:00Z"))
	child := NewChild(SpanID("c"), root, SpanKindTurn, ts("2026-05-16T00:00:01Z"))
	if child.ParentSpanID == nil || *child.ParentSpanID != "r" {
		t.Fatalf("parent=%+v", child.ParentSpanID)
	}
	if child.SessionID != sid("s") {
		t.Fatalf("session=%v", child.SessionID)
	}
	if child.Status.Kind != SpanStatusKindOk {
		t.Fatalf("status=%v", child.Status)
	}
}

func TestSpanBaseFinish(t *testing.T) {
	b := NewRoot("r", sid("s"), tid("t"), SpanKindTurn, ts("2026-05-16T00:00:00Z"))
	b.Finish(ts("2026-05-16T00:00:02Z"), NewStatusError("bad"), 2000)
	if b.DurationMs != 2000 || b.Status.Kind != SpanStatusKindError || b.Status.Message != "bad" {
		t.Fatalf("finish: %+v", b)
	}
}

// ─── Rule: GetSessionMetrics returns nil when nothing recorded ──────────────

func TestGetSessionMetricsReturnsNilWhenEmpty(t *testing.T) {
	obs := NewInMemoryObservabilityProvider()
	m, err := obs.GetSessionMetrics(context.Background(), sid("missing"))
	if err != nil {
		t.Fatal(err)
	}
	if m != nil {
		t.Fatalf("want nil, got %+v", m)
	}
}
